[fx] Add DCE pass

jfix71 opened the pull request:
Summary: DCE will reverse iterate over the graph looking for nodes without users and delete them.

It will skip over unused impure nodes:
- placeholders (since this affects the signature of the method) 
- outputs (also affects method signature, and additionally never have users but we want to keep them 🙂 )
- functions explicitly listed as side-effectful
- modules which have `_is_impure = True`

Test Plan: Added unit tests

Differential Revision: D26602212

jamesr66a commented on torch/fx/passes/dce.py:
```diff
@@ -0,0 +1,13 @@
+    # removed node.
+    for node in reversed(m.graph.nodes):
+        if node.op not in {"placeholder", "output"} and len(node.users) == 0:
+            m.graph.erase_node(node)
```
This could probably be a method on `Graph`. That way we can also keep the semantics in one place. In particular, I'm thinking about if in the future we have more side-effectful nodes we need to preserve other than placeholders and outputs (e.g. asserts)

jamesr66a commented on torch/fx/graph.py:
```diff
@@ -952,6 +952,20 @@ def check_arg(arg : Node, n : Optional[Node] = None) -> None:
+        # input to that node have an updated user count that no longer reflects
+        # the removed node.
+        for node in reversed(self.nodes):
+            if node.op not in {"placeholder", "output"} and len(node.users) == 0:
```
Let's move this condition out into a method `Node.is_impure`. That way we can have a single, overridable definition for impurity of nodes that other passes can utilize as well

jfix71 replied:
Cool, so I've added it -- but I also added another quirk here in the latest update. Specifically, some other nodes may have side effects which means we should not remove them even if they have no users.

This may be universal (i.e. regardless of use case), for example for `call_function` to `torch._assert`, and so could be added to `is_impure()` in a straightforward way.

But it may also be use case dependent -- for example if we have some special Module that we don't trace into but we know it has its own asserts or raises exceptions, and so should not be eliminated.

I added more special handling for these cases for now inside `eliminate_dead_code()` directly, with extra args passed in to handle them, but would be great to get your feedback on if it's handled in the best way -- e.g. perhaps we should move those to `Node.is_impure()`? Or perhaps something else.

jamesr66a commented on torch/fx/graph.py:
```diff
@@ -952,6 +952,20 @@ def check_arg(arg : Node, n : Optional[Node] = None) -> None:
                             raise RuntimeError(f'Node {node} target {node.target} references nonexistent attribute '
                                                f'{atom} of {seen_qualname}')
 
+    def run_dce(self):
```
I'd prefer a name like `eliminate_dead_code`. Not sure if our general userbase will be familiar with the DCE initialism

jamesr66a commented on torch/fx/graph.py:
```diff
@@ -952,6 +952,20 @@ def check_arg(arg : Node, n : Optional[Node] = None) -> None:
+        Remove all dead code from the graph, based on each node's number of users.
+        """
+        # Lint the graph first to make sure its topologically sorted, otherwise
+        # DCE below will not behave as expected.
```
Mention this invariant in the docstring

jamesr66a commented on torch/fx/graph.py:
```diff
@@ -952,6 +952,20 @@ def check_arg(arg : Node, n : Optional[Node] = None) -> None:
 
+    def run_dce(self):
+        """
+        Remove all dead code from the graph, based on each node's number of users.
```
Can you format this according to the Google Python docstring style: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

Would probably also be nice to have a quick example in the docstring

jamesr66a reviewed (commented):
Few more comments then I think this will be good

jamesr66a commented on torch/fx/node.py:
```diff
@@ -393,6 +393,16 @@ def maybe_replace_node(n : Node) -> Node:
+
+            bool: If the op is impure or not.
+        """
+        return self.op in {"placeholder", "output"}
```
I recommend doing something like (after https://github.com/pytorch/pytorch/pull/52358 is landed):

```
# At global scope
_side_effectful_modules : Set[Type] = {}
_side_effectful_functions : Set[Callable] = {}

<...>

    def is_impure(self):
        if self.op in {"placeholder", "output"}:
          return True

        if self.op == 'call_function':
          return self.target in _side_effectful_functions

        if self.graph.owning_module and self.op == 'call_module':
            # This could be cleaned up if that PR changes `has_submodule` to
            # `get_submodule`
            assert self.graph.owning_module.has_submodule(node.target)
            mod = self.graph.owning_module
            for atom in self.target.split('.'):
                mod = getattr(md, atom)

            # Check for known side-effectful Modules.
            # Note we could also delete this if we just mark
            # those modules as `_is_impure`
            if type(mod) in _side_effectful_modules:
                return True

            # Check if the Module has an `_is_impure` attribute
            # If it's true, let's take their word for it
            return getattr(mod, '_is_impure', False)

        return False

```

jamesr66a replied:
For checking if functions have side effects, it might also be possible to pull out the TorchScript schema for a given PyTorch op and check if there is aliasing. e.g. I pull out the schema for an op here: https://github.com/pytorch/pytorch/pull/49775/files#diff-03652431ccd14a2e043861738619b0b73977b3ce531f28b10946b1aacdc7456fR90 and querying the aliasing info shouldn't be too hard.

Having a hardcoded registry at first might be OK though, we can explore adding this afterward

jfix71 commented:
Updated accordingly, but not stacked on #52358 yet, so for now still passing the module in to retrieve the type.

jfix71 committed:
[fx] Add DCE pass (#52658)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/52658

DCE will reverse iterate over the graph looking for nodes without users and delete them. It will skip over unused placeholders (since this affects the signature of the method) and outputs (which never have users but we want to keep them :) )

Test Plan: Added unit tests

Reviewed By: jamesr66a, khabinov, chenccfb

Differential Revision: D26602212

fbshipit-source-id: 0db10cc91f2c8e6153cccd9bb027dcbc4df2e941
